### ISSUE TITLE
Add concurrent batching to the `file` sink

### ISSUE THREAD

**Issue by fpytloun** [vectordotdev/vector#20394]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

I can see 100% utilization on file sink which then applies backpressure and slows-down whole pipeline. I am using tmpfs so disk is not a bottleneck but high cardinal partitioning could be. It seems that file sink is not batching concurrently and therefore applying backpressure quickly (especially with gzip compression).

### Configuration

```text
[sinks.out_kafka_access_file]
      type = "file"
      #inputs = ["throttle_kafka_access_tenant"]
      inputs = ["remap_kafka_access"]
      compression = "gzip"
      encoding.except_fields = ["_index", "_topic", "_topic_template", "_partition", "_offset", "_throttle_key", "_hash", "_alert", "_keep", "_sd", "_source", "_syslog_severity", "_file_suffix", 'kubernetes.labels."pod-template-hash"', "@source_type", "@metadata"]
      encoding.codec = "json"
      framing.method = "newline_delimited"
      # ._file_suffix = to_int(to_int(now()) / 300)
      path = "/var/lib/vector/s3sync/out_kafka_access_file/topics/{{ _topic }}/year=%Y/month=%m/day=%d/hour=%H/${HOSTNAME}.pa2-par-gc-int-ves-io_{{ _file_suffix }}.json.gz"
      idle_timeout_secs = 30
      buffer.type = "memory"
      buffer.max_events = 3000    # default 500 with memory buffer
```

### Version

0.37.0

**Comment by jszwedko** [comment]
Hi @fpytloun ,

Thanks for filing this! I think I'm missing what the specific request is though. Is it to generally improve the throughput of the `file` sink? Or to make some specific modification to it? I can think of a few modifications that could be made:

- Add batching (right now, no batching occurs) (related: https://github.com/vectordotdev/vector/issues/2174)
- Process batches concurrently (requires batching first)

**Comment by fpytloun** [comment]
Hello @jszwedko, so I was thinking of this as a bug because this component which should be very simple can easily apply backpressure (even when using tmpfs) and limit more complex components (Kafka, Elasticsearch, Clickhouse, etc.) that can however handle much higher throughput.

I think batching and concurrency is what would fix it.

**Comment by jszwedko** [comment]
> Hello @jszwedko, so I was thinking of this as a bug because this component which should be very simple can easily apply backpressure (even when using tmpfs) and limit more complex components (Kafka, Elasticsearch, Clickhouse, etc.) that can however handle much higher throughput.
> 
> I think batching and concurrency is what would fix it and I currently don't know of any workaround. Possibly use aws_s3 now which will consume lot of memory due to in-memory batching but probably same amount as I am already using tmpfs with file sink.

Makes sense, thanks for confirming. I'll adjust this issue to be about adding concurrent batching to the sink.

**Comment by scMarkus** [comment]
@jszwedko I have come across this since I would like to improve the file sinks throughput. Would it be reasonable to add batching to this sink similar to other sinks? And if so what would be a good sink to look into the implementation and steal some code from? http maybe?

**Comment by pront** [comment]
> [@jszwedko](https://github.com/jszwedko) I have come across this since I would like to improve the file sinks throughput. Would it be reasonable to add batching to this sink similar to other sinks? And if so what would be a good sink to look into the implementation and steal some code from? http maybe?

Hi @scMarkus, we would like the same UX with other components. You can look at the HTTP partition keys to begin with and take it from there. Also, maybe @jszwedko has more to add here.

**Comment by jszwedko** [comment]
Not too much more to add. I think the throughput would improve with batching. I agree that the `http` sink could be a good one to take a look at as an example implementation of batching.

**Comment by scMarkus** [comment]
First of all I am very sorry for not showing any meaningful progression after all this time.
Still I am quite interested in any improvements made. @jszwedko could you mind taking a look at [this](https://github.com/vectordotdev/vector/pull/25509) new pull request

**Comment by pront** [comment]
> First of all I am very sorry for not showing any meaningful progression after all this time. Still I am quite interested in any improvements made. [@jszwedko](https://github.com/jszwedko) could you mind taking a look at [this](https://github.com/vectordotdev/vector/pull/25509) new pull request

Hi @scMarkus that PR needs some work. Are you interested in cleaning it up? If so feel free to take it over and ping @vectordotdev/vector if you need any further help.

**Comment by scMarkus** [comment]
@pront i created another pull requests [here](https://github.com/vectordotdev/vector/pull/26081) so we have something to look at

there are 2 things I am quite unhappy with:
- I rebased the original commit and now it shows as mine in the git history
- since `sink.partition_event()` function is not used anymore tests like [this](https://github.com/smartclip/vector/blob/aefa907a55ed9c78ea8dbc5d6a3d7da69f05eccb/src/sinks/file/mod.rs#L1329-L1346) do not work. For now I have commented them out but I do no feel confident in this being the right way to handle it.

**Comment by scMarkus** [comment]
@vectordotdev/vector
I have progressed the [pullrequest](https://github.com/vectordotdev/vector/pull/26081) to a point where at least codex is not complaining anymore. How would we proceed?

**Comment by scMarkus** [comment]
@pront i noticed that you added `meta: awaiting author` a few days ago. after that I updated the branch again. I was not aware that this removes said label. sorry
